### PR TITLE
add more tests and doc (3) - last one

### PR DESCRIPTION
**Pull request by anporumb**

### CODE REVIEW

File: doc/macro_utils_design.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
+macro utils design
+=======
+
+Macro utils is a set of C useful macros.
+
+This document explains how it was build and what every macro does. IN supporting understand rather than code accuracy, some macros will be chopped to deal with a reasonable number of arguments rather than their implementation that usually supports above 120 arguments.
```
Comment (dcristoloveanu) [inline doc/macro_utils_design.md:6]: >IN [](start = 67, length = 2)

In

Comment (dcristoloveanu) [inline doc/macro_utils_design.md:6]: >understand [](start = 81, length = 10)

understandability (or understanding the code)

File: doc/macro_utils_design.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (dcristoloveanu) [inline doc/macro_utils_design.md:240]: Lots of empty lines that can be removed

File: tests/main.c [hunk 3]
```diff
@@ -5,49 +5,69 @@
...
+    //POOR_MANS_ASSERT(result == 0);
+    //
+    //result = run_mu_define_local_enum_tests();
+    //POOR_MANS_ASSERT(result == 0);
+    //
+    //result = run_mu_define_enum_2_tests();
+    //POOR_MANS_ASSERT(result == 0);
+    //
+    //result = run_mu_pri_enum_tests();
+    //POOR_MANS_ASSERT(result == 0);
```
Comment (dcristoloveanu) [inline tests/main.c:70]: Suspicion of these being erroneously commented

File: tests/mu_count_1_or_more_arg_test.c [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        114,
+        115,
+        116,
+        117,
+        118,
+        119,
+        120,
+        121,
+        122,
+            123) == 123);
```
Comment (dcristoloveanu) [inline tests/mu_count_1_or_more_arg_test.c:147]: >    [](start = 8, length = 3)

Nitpick: extra indent.

File: doc/macro_utils_design.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+Obviously `d` is not the expected macro expansion at this moment. So the macro has some limitations, it can count only 1...3 arguments. (it can count 1..n-1, where n is coming from `MU_THE_NTH_ARG`)
+
+So what happens if the macro is called with 0 arguments?
+MU_COUNT_1_OR_MORE_ARG() => MU_THE_NTH_ARG( , 3, 2, 1) => 1
+                                                    ^
+                                                    |----- this is 4th argument
+                                                
+Here we see how the macro does not produce 0, and that is precisely why the macro is called MU_COUNT_**1**_OR_MORE_ARG. 
+
+Next question is : how to produce 0 as a count result? Not easy or immediately stright forward. More macros are needed for that.
```
Comment (dcristoloveanu) [inline doc/macro_utils_design.md:64]: >stright [](start = 79, length = 7)

straight

File: doc/macro_utils_design.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+### MU_HAS_COMMA(...)
+
+`MU_HAS_COMMA` expands to 0 when the number of its arguments is 0 or 1. If the number of arguments is greater or equal to 2 then it expands to 0. Here' some examples
```
Comment (dcristoloveanu) [inline doc/macro_utils_design.md:68]: >Here' [](start = 146, length = 5)

Here are some

File: doc/macro_utils_design.md (hunk 1 above)
Comment (anporumb) [reply doc/macro_utils_design.md:6]: In, in.

---
In reply to: [464137885](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464137885) [](ancestors = 464137885)

Comment (anporumb) [reply doc/macro_utils_design.md:6]: added ability

---
In reply to: [464137916](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464137916) [](ancestors = 464137916)

File: doc/macro_utils_design.md (hunk 2 above)
Comment (anporumb) [reply doc/macro_utils_design.md:267]: Removed

---
In reply to: [464138072](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464138072) [](ancestors = 464138072)

File: tests/main.c (hunk 3 above)
Comment (anporumb) [reply tests/main.c]: Absolutely. They are now (iteration 6) uncommented. I believe you've looked over a draft, but even so, your comments are valid.

---
In reply to: [464138282](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464138282) [](ancestors = 464138282)

File: tests/mu_count_1_or_more_arg_test.c (hunk 4 above)
Comment (anporumb) [reply tests/mu_count_1_or_more_arg_test.c:147]: Unintentionally indented - now intentionally unintended. 

---
In reply to: [464138344](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464138344) [](ancestors = 464138344)

File: doc/macro_utils_design.md (hunk 5 above)
Comment (anporumb) [reply doc/macro_utils_design.md:64]: straightened.

---
In reply to: [464139262](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464139262) [](ancestors = 464139262)

File: doc/macro_utils_design.md (hunk 6 above)
Comment (anporumb) [reply doc/macro_utils_design.md:68]: grammarified.

---
In reply to: [464139410](https://github.com/Azure/azure-macro-utils-c/pull/42#discussion_r464139410) [](ancestors = 464139410)

**Pull request merged**